File: src/states/create.cpp
#include "states/create.h"

static StateFn new_create_state() {
    StateFn s;
    s.event_handler = create_fn::event_handler;
    s.draw = create_fn::draw;
    return s;
}

void add_cell(StateRec& rec, const size_t x, const size_t y) {
    if (rec.current_line.empty() || (rec.current_line.back().x == x && rec.current_line.back().y)) {
        return;
    }
    size_t index = get_index(x, y, rec.grid);
    if (((rec.drag == CLEAR && rec.grid.cells[index] == rec.current_fill) || (rec.drag == DRAW && rec.grid.cells[index] != rec.current_fill))) {
        Point p;
        p.x = x;
        p.y = y;
        rec.current_line.push_back(p);
    }
}

bool create_fn::event_handler(StateFn& fn, StateRec& rec) {
    if (IsKeyPressed(KEY_ESCAPE)) {
        fn = HOME_STATE;
    }
    if (IsMouseButtonPressed(MOUSE_BUTTON_LEFT) && rec.current_fill == Empty) {
        Vector2 mouse_pos = GetMousePosition();
        size_t x = find_cell_x(mouse_pos.x, rec);
        size_t y = find_cell_y(mouse_pos.y, rec);
        rec.current_fill = Solid;
        if (x < rec.grid.n_cols && y < rec.grid.n_rows) {
            if (rec.grid.cells[y][x] == Solid) {
                rec.drag = CLEAR;
            }
            if (rec.grid.cells[y][x] != Cross) {
                Point p;
                p.x = x;
                p.y = y;
                rec.current_line.push_back(p);
            }
        }
    } else if (IsMouseButtonDown(MOUSE_BUTTON_LEFT) && rec.current_fill == Solid) {
        Vector2 mouse_pos = GetMousePosition();
        size_t x = find_cell_x(mouse_pos.x, rec);
        size_t y = find_cell_y(mouse_pos.y, rec);
        if (x < rec.grid.n_cols && y < rec.grid.n_rows) {
            if (rec.grid.cells[y][x] != Cross) {
                add_cell(rec, x, y);
            }
        }
    }

    if (IsMouseButtonPressed(MOUSE_BUTTON_MIDDLE) && rec.current_fill == Empty) {
        Vector2 mouse_pos = GetMousePosition();
        size_t x = find_cell_x(mouse_pos.x, rec);
        size_t y = find_cell_y(mouse_pos.y, rec);
        rec.current_fill = Note;
        if (x < rec.grid.n_cols && y < rec.grid.n_rows) {
            if (rec.grid.cells[y][x] == Note) {
                rec.drag = CLEAR;
            }
            if (rec.grid.cells[y][x] == Empty || rec.grid.cells[y][x] == Note) {
                Point p;
                p.x = x;
                p.y = y;
                rec.current_line.push_back(p);
            }
        }
    } else if (IsMouseButtonDown(MOUSE_BUTTON_MIDDLE) && rec.current_fill == Note) {
        Vector2 mouse_pos = GetMousePosition();
        size_t x = find_cell_x(mouse_pos.x, rec);
        size_t y = find_cell_y(mouse_pos.y, rec);
        if (x < rec.grid.n_cols && y < rec.grid.n_rows) {
            if (rec.grid.cells[y][x] == Empty || rec.grid.cells[y][x] == Note) {
                add_cell(rec, x, y);
            }
        }
    }


    if (IsMouseButtonPressed(MOUSE_BUTTON_RIGHT) && rec.current_fill == Empty) {
        Vector2 mouse_pos = GetMousePosition();
        size_t x = find_cell_x(mouse_pos.x, rec);
        size_t y = find_cell_y(mouse_pos.y, rec);
        rec.current_fill = Cross;
        if (x < rec.grid.n_cols && y < rec.grid.n_rows) {
            if (rec.grid.cells[y][x] == Cross) {
                rec.drag = CLEAR;
            }
            if (rec.grid.cells[y][x] != Solid) {
                Point p;
                p.x = x;
                p.y = y;
                rec.current_line.push_back(p);
            }
        }
    } else if (IsMouseButtonDown(MOUSE_BUTTON_RIGHT) && rec.current_fill == Cross) {
        Vector2 mouse_pos = GetMousePosition();
        size_t x = find_cell_x(mouse_pos.x, rec);
        size_t y = find_cell_y(mouse_pos.y, rec);
        if (x < rec.grid.n_cols && y < rec.grid.n_rows) {
            if (rec.grid.cells[y][x] != Solid) {
                add_cell(rec, x, y);
            }
        }
    }

    if ((IsMouseButtonReleased(MOUSE_BUTTON_LEFT) && rec.current_fill == Solid) || (IsMouseButtonReleased(MOUSE_BUTTON_MIDDLE) && rec.current_fill == Note) || (IsMouseButtonReleased(MOUSE_BUTTON_RIGHT) && rec.current_fill == Cross)) {
        for (auto& p: rec.current_line) {
            set_cell(rec, p);
        }
        rec.drag = DRAW;
        rec.current_fill = Empty;
        rec.current_line.clear();
    }
    return true;
}

static char SAVE_TEXT[] = "Press S to save the current Nonogram";
static int SAVE_SIZE = 20;


static Style style = {
        .foreground = FOREGROUND,
        .background = BACKGROUND,
        .thickness = 2.0,
        .radius = 4.0
};

void create_fn::draw(StateRec& rec) {
    BeginDrawing();
    {
        ClearBackground(BACKGROUND);
        DrawFPS(0,0);
        DrawText(SAVE_TEXT, (GetScreenWidth() - MeasureText(SAVE_TEXT, SAVE_SIZE)) / 2, rec.bounds.top - SAVE_SIZE * 2, SAVE_SIZE, FILL);
        color_cells(rec, style);
        draw_grid(rec, style.foreground);
    }
    EndDrawing();
}

File: create.go
package main

import (
	rl "github.com/gen2brain/raylib-go/raylib"
)

var CREATE = Create{}

type Create struct {
	rec *State
}

func (c *Create) setRec(rec *State) {
	c.rec = rec
}

func (c *Create) getRec() *State {
	return c.rec
}

func (c *Create) handleEvents() bool {
	if rl.IsKeyPressed(rl.KeyEscape) {
		c.rec.fn = &HOME
	}
	solidTest := c.checkAndAddAt(rl.MouseButtonLeft, Solid, Cross)
	crossTest := c.checkAndAddAt(rl.MouseButtonRight, Cross, Solid)
	noteTest := c.checkAndAddAt(rl.MouseButtonMiddle, Note, Solid, Cross)

	if solidTest() || crossTest() || noteTest() {
		for _, p := range c.rec.line {
			c.rec.grid.setCell(&p, c.rec.fill)
		}
		c.rec.drag = Draw
		c.rec.fill = Empty
		c.rec.line = c.rec.line[:0]
	}
	return true
}

func (c *Create) draw() {
	rl.BeginDrawing()
	{
		rl.ClearBackground(c.rec.style.background)
		rl.DrawFPS(0, 0)
		c.rec.grid.colorCells(c.rec.cellSize, &c.rec.bounds, &c.rec.style)
		c.rec.colorTempCells()
		c.rec.grid.drawGrid(c.rec.cellSize, &c.rec.bounds, c.rec.style.foreground)
	}
	rl.EndDrawing()
}

func (c *Create) checkAndAddAt(mouseButton rl.MouseButton, fill FillType, protected ...FillType) func() bool {
	testFn := func() bool {
		return rl.IsMouseButtonReleased(mouseButton) && c.rec.fill == fill
	}
	if (rl.IsMouseButtonPressed(mouseButton) && c.rec.fill == Empty) || (rl.IsMouseButtonDown(mouseButton) && c.rec.fill == fill) {
		p := c.rec.findCellXY(rl.GetMousePosition())
		if p.x < c.rec.grid.nCols && p.y < c.rec.grid.nRows {
			index := c.rec.grid.getIndex(&p)
			if c.rec.grid.cells[index] == fill && c.rec.fill == Empty {
				c.rec.drag = Clear
			}
			c.rec.fill = fill
			under := c.rec.grid.cells[index]
			for _, prot := range protected {
				if under == prot {
					return testFn
				}
			}
			c.rec.addTempCell(p)
		}
	}
	return testFn
}
